Python/bltinmodule.c

```diff
@@ -2628,36 +2645,86 @@ zip_next(zipobject *lz)
             item = (*Py_TYPE(it)->tp_iternext)(it);
             if (item == NULL) {
                 Py_DECREF(result);
+                if (lz->strict) {
+                    goto check;
+                }
                 return NULL;
             }
             PyTuple_SET_ITEM(result, i, item);
         }
     }
     return result;
+check:
+    if (PyErr_Occurred() && !PyErr_ExceptionMatches(PyExc_StopIteration)) {
```

Comment:
The code in the error label is non-trivial. You may add comments. Here I propose:

```
// next() on argument i raised an exception (not StopIteration)
```